File: dotnet/picocontainer/src/Defaults/ImplementationHidingComponentAdapter.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using System.Collections;
using System.Reflection;
using System.Runtime.Remoting.Contexts;	
using System.Runtime.Remoting.Activation;
using System.Runtime.Remoting.Messaging;
using System.Runtime.Remoting.Proxies;
using System.Runtime.Remoting;
using System.Runtime.Remoting.Services;

namespace PicoContainer.Defaults {
  public class ImplementationHidingComponentAdapter : DecoratingComponentAdapter {

    private readonly InterfaceFinder interfaceFinder = new InterfaceFinder();
    private bool strict;

    public ImplementationHidingComponentAdapter(IComponentAdapter theDelegate, bool strict) : base(theDelegate){
      this.strict = strict;
    }

    public ImplementationHidingComponentAdapter(IComponentAdapter theDelegate) : base(theDelegate){
    }

    public override object ComponentInstance {

      get {
        Type[] interfaces;
        Type delegateType = this.Delegate.ComponentKey as Type;
        if (delegateType != null && delegateType.IsInterface) {
          interfaces = new Type[] {delegateType};
        } else {
          interfaces = interfaceFinder.GetInterfaces(Delegate.ComponentImplementation);
        }
        if (interfaces.Length == 0) {
          if (strict) {
            throw new PicoIntrospectionException("Can't hide implementation for " + Delegate.ComponentImplementation.Name + ". It doesn't implement any interfaces.");
          } else {
            return Delegate.ComponentInstance;
          }
        }

        return new DelegatingInvocationHandler(this, interfaces).GetTransparentProxy();
      }
    }

    private object DelegatedComponentInstance {
      get {
        return base.ComponentInstance;
      }
    }


    private class DelegatingInvocationHandler : RealProxy, IRemotingTypeInfo, ISwappable{ 
      private ImplementationHidingComponentAdapter adapter;
      private object delegatedInstance;
      private ArrayList interfaces;
      private MethodInfo hotSwapMethod = null;
      public DelegatingInvocationHandler(ImplementationHidingComponentAdapter adapter, Type[] interfaces) : base(typeof(ISwappable)) {
        this.adapter = adapter;
        this.interfaces = new ArrayList(interfaces);
      }
      
      
      // C# implementation problem, there is for as far as I know no System.identityHashCode available for c#
      // to perform the GetHashCode() method.
      public override IMessage Invoke(IMessage msg) {
        IMethodCallMessage call = (IMethodCallMessage)msg;
        IConstructionCallMessage ctor = call as IConstructionCallMessage;
        IMethodMessage m = msg as IMethodMessage;
        bool isHostSwapCall = call.MethodName == "HotSwap";
        if (delegatedInstance == null) {
          delegatedInstance = adapter.DelegatedComponentInstance;
          if (isHostSwapCall)
          {
            hotSwapMethod = delegatedInstance.GetType().GetMethod("HotSwap",new Type[] {m.Args[0].GetType()});
          }
        }
        if (isHostSwapCall) {
          if ( hotSwapMethod == null) {
            return new ReturnMessage(m.MethodBase.Invoke(this,BindingFlags.Public,null,m.Args,null),new object[]{},0,call.LogicalCallContext,call);
          } else
          {
            return new ReturnMessage(hotSwapMethod.Invoke(delegatedInstance,BindingFlags.Public,null,m.Args,null),new object[]{},0,call.LogicalCallContext,call);
          }
        } else {
          return new ReturnMessage(m.MethodBase.Invoke(delegatedInstance,BindingFlags.Public,null,m.Args,null),new object[]{},0,call.LogicalCallContext,call);
        }
      }

      public object HotSwap(object newSubject) {
        object result = delegatedInstance;
        delegatedInstance = newSubject;
        return result;
      }

      string typeName;
      public string TypeName {
        get {
          return typeName;
        }
        set {
          typeName = value;
        }
      }

      public bool CanCastTo(Type testType,object obj) {
        return interfaces.Contains(testType) || testType == typeof(ISwappable);
      }
    }  
  }
}

File: java/picocontainer/src/java/org/picocontainer/defaults/ImplementationHidingComponentAdapter.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Original code by                                                          *
 *****************************************************************************/
package org.picocontainer.defaults;

import org.picocontainer.ComponentAdapter;
import org.picocontainer.PicoInitializationException;
import org.picocontainer.PicoIntrospectionException;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

/**
 * This component adapter makes it possible to hide the implementation
 * of a real subject (behind a proxy).
 * If the key of the component is of type {@link java.lang.Class} and that class represents an interface, the proxy
 * will only implement the interface represented by that Class. Otherwise (if the key is
 * something else), the proxy will implement all the interfaces of the underlying subject.
 * In any case, the proxy will also implement
 * {@link Swappable}, making it possible to swap out the underlying
 * subject at runtime.
 * <p>
 * <em>
 * Note that this class doesn't cache instances. If you want caching,
 * use a {@link CachingComponentAdapter} around this one.
 * </em>
 * 
 * @author Paul Hammant
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public class ImplementationHidingComponentAdapter extends DecoratingComponentAdapter {
    private static Method hotswap;
    static {
        try {
            hotswap = Swappable.class.getMethod("hotswap", new Class[]{Object.class});
        } catch (NoSuchMethodException e) {
            throw new InternalError();
        }
    }

    private final boolean strict;

    /**
     * Alternative constructor allowing to set interface-only strictness.
     * @param delegate the delegate adapter
     * @param strict true if the adapter should only accept classes that are hideable behind interfaces.
     * If false, a non-implementation hidden instance will be created instead of throwing an exception.
     */
    public ImplementationHidingComponentAdapter(ComponentAdapter delegate, boolean strict) {
        super(delegate);
        this.strict = strict;
    }


    /**
     * Creates a strict ImplementationHidingComponentAdapter that will throw an exception
     * when trying to instantiate a class that doesn't implement any interfaces.
     * @param delegate the delegate adapter
     */
    public ImplementationHidingComponentAdapter(ComponentAdapter delegate) {
        this(delegate, true);
    }

    public Object getComponentInstance()
            throws PicoInitializationException, PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {

        Class[] interfaces;
        if(getDelegate().getComponentKey() instanceof Class && ((Class)getDelegate().getComponentKey()).isInterface()) {
            interfaces = new Class[] {(Class) getDelegate().getComponentKey()};
        } else {
            interfaces = ClassHierarchyIntrospector.getAllInterfaces(getDelegate().getComponentImplementation());
        }
        if (interfaces.length == 0) {
            if(strict) {
                throw new PicoIntrospectionException("Can't hide implementation for " + getDelegate().getComponentImplementation().getName() + ". It doesn't implement any interfaces.");
            } else {
                return getDelegate().getComponentInstance();
            }
        }
        Class[] swappableAugmentedInterfaces = new Class[interfaces.length + 1];
        swappableAugmentedInterfaces[interfaces.length] = Swappable.class;
        System.arraycopy(interfaces, 0, swappableAugmentedInterfaces, 0, interfaces.length);
        final DelegatingInvocationHandler delegatingInvocationHandler = new DelegatingInvocationHandler(this);
        return Proxy.newProxyInstance(
                getClass().getClassLoader(),
                swappableAugmentedInterfaces,
                delegatingInvocationHandler);
    }

    private Object getDelegatedComponentInstance() {
        return super.getComponentInstance();
    }

    private class DelegatingInvocationHandler implements InvocationHandler {
        private final ImplementationHidingComponentAdapter adapter;
        private Object delegatedInstance;

        public DelegatingInvocationHandler(ImplementationHidingComponentAdapter adapter) {
            this.adapter = adapter;
        }

        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            Class declaringClass = method.getDeclaringClass();
            if (declaringClass.equals(Object.class)) {
                if (method.equals(ClassHierarchyIntrospector.hashCode)) {
                    // Return the hashCode of ourself, as Proxy.newProxyInstance() may
                    // return cached proxies. We want a unique hashCode for each created proxy!
                    return new Integer(System.identityHashCode(this));
                }
                if (method.equals(ClassHierarchyIntrospector.equals)) {
                    return new Boolean(proxy == args[0]);
                }
                // If it's any other method defined by Object, call on ourself.
                return method.invoke(this, args);
            } else if (hotswap.equals(method)) {
                return hotswap(args[0]);
            } else {
                if (delegatedInstance == null) {
                    delegatedInstance = adapter.getDelegatedComponentInstance();
                }
                return method.invoke(delegatedInstance, args);
            }
        }

        private Object hotswap(Object newSubject) {
            Object result = delegatedInstance;
            delegatedInstance = newSubject;
            return result;
        }
    }
}
